### PR TITLE
Bug 2021205: fix git url change validation

### PR DESCRIPTION
**Pull request by abhinandan13jan**
# Addresses 
https://issues.redhat.com/browse/OCPBUGSM-36949
https://bugzilla.redhat.com/show_bug.cgi

# Issue
Reads `gitType` instead of `detectedGitType`

# Screenshots
url-change-validation

# Tests
no change

# Browser conformance
Chrome / Firefox

### CODE REVIEW

**Comment by karthikjeeyar** [comment]
When the user fixes the incorrect url, the `Git type` should not be shown. We need to reset the showGitType variable when a valid git type is detected


image

**Comment by abhinandan13jan** [comment]
> When the user fixes the incorrect url, the `Git type` should not be shown. We need to reset the showGitType variable when a valid git type is detected
> 
> image

Fixed
git-url-valid

**Review by jerolimov — CHANGES_REQUESTED** [review]
In the old version, the user must select the Git provider if the URL detection doesn't work.

This single case is fixed in the new version, but it breaks the option to select a provider .. and maybe switch it after that. Checkout this recoding:

Peek 2022-01-05 22-24

/hold

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/GitSection.tsx:247]: Does this also if we compare it to "not unsure" (aka sure) instead of list all (currently) known Git provider?

You can optional add the && showGitTypes so that `setFieldValue` is not called if it's already false.

```suggestion
      } else if (detectedGitType !== GitTypes.unsure && values.git.showGitType) {
```

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/GitSection.tsx:260]: I'm not sure, but is this change really needed? Doesn't it make sense to use the selected type if we show this dropdown option?

**Comment by abhinandan13jan** [comment]
> In the old version, the user must select the Git provider if the URL detection doesn't work.
> 
> This single case is fixed in the new version, but it breaks the option to select a provider .. and maybe switch it after that. Checkout this recoding:
> 
> Peek 2022-01-05 22-24
> 
> /hold

I updated
git-url-change

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/import/git/GitSection.tsx]: Updated

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/import/git/GitSection.tsx]: i am now using a cleaner approach to the problem. Updated

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (jerolimov) [inline frontend/packages/dev-console/src/components/import/git/GitSection.tsx:253]:
```suggestion
      if (gitType !== values.git.type) {
        setFieldTouched('git.type', false, false);
        setFieldValue('git.type', gitType, true);
      }
```

With this change:

fixed

**Review by jerolimov — COMMENTED** [review]
Hi @abhinandan13jan, cc @karthikjeeyar,

the reason my review takes so long was that I'm not happy that `handleGitUrlChange` was now called twice when switching from an invalid to a valid git URL.

duplicate-github-calls

This happen because `const handleGitUrlChange = React.useCallback(...` updates the git type in line 252. The `useEffect` in line 421 then calls the same `handleGitUrlChange` again with a debounce.

I think we could fix it by cleaning the touched info when we automatically override the `git.type` in line 251. See suggestion below.

Did you see any reasons, not to do this? IMHO this works only because the `useEffect` has a (for me not really understandable) condition for different "touched" fields. Btw: I'm not a fan of this solution and that the InputField onChange, this callback, and the effect works so closely together, but that is something we should maybe cleanup in the future.

**Comment by jerolimov** [comment]
@abhinandan13jan @karthikjeeyar I replace Karthiks lgtm with my approval so that you can updates this on Monday without my additional review. But I will do if needed.

I hope that's fine for you both. If you both agree that we can not apply my recommendation, Karthik can just re-add his lgtm. :smile: 

/lgtm cancel
/approve

**Comment by karthikjeeyar** [comment]
@abhinandan13jan I agree with Christoph's concern, we should try avoid multiple calls for git detection. Please update the PR. 

Thanks @jerolimov for the review and highlighting the issue with multiple network calls, which I overlooked. However I feel as reviewers we should not give lgtm / approve unless we are satisfied with the changes.

File: frontend/packages/dev-console/src/components/import/git/GitSection.tsx
Comment (abhinandan13jan) [reply frontend/packages/dev-console/src/components/import/git/GitSection.tsx]: Thanks for the feedback, I updated accordingly :)

**Pull request merged**
